### ISSUE TITLE
window explorer (and terminal) hangs when I edit the address bar in a file picker

### ISSUE THREAD

**Issue by KamenRevice** [microsoft/terminal#17028]
### Windows Terminal version

1.19.10573.0

### Windows build number

 10.0.19045

### Other Software

Windows explore
![image](https://github.com/microsoft/terminal/assets/36665957/3dadd74a-3519-4782-bab8-fc16d4912582)

### Steps to reproduce

In the icon option for editing a single profile, after using the browse option and opening window explorer, it is not possible to enter the path to the file directly in the address bar, it is not possible to enter any characters, which causes the entire Windows terminal to go down, and the only way to close the program is to close it through task management.
please look the gif animation

### Expected Behavior

![image](https://github.com/microsoft/terminal/assets/36665957/d759f07b-c5e5-4f75-817c-94a7bfee3d49)

### Actual Behavior

software is dead

**Comment by zadjii-msft** [comment]
Hmm. We once had a similar report in #8957, but that was when users were running the Terminal elevated, and was also fixed in like, 1.7.

There was also #9984, which sounds almost exactly the same, but OP disappeared on that one.

I also can't repro this here on Windows 11. 

Filing /feedback with the Feedback hub (but with the area set to like, "Files and Online Storage > File Explorer") may work to get a trace:

![image](https://github.com/microsoft/terminal/assets/18356694/87f315e0-6b06-4b6f-ba44-72b8b125e68c)

**Comment by KamenRevice** [comment]
Here's a link to the feedback recorded in the feedback [center](https://aka.ms/AAq4985)
Does this have anything to do with the network driver[ clouddriver2](https://www.clouddrive2.com/) that I installed

**Comment by zadjii-msft** [comment]
Huh. Interestingly enough, it looks like you're the only one who's hit this? 

* Apparently I can't promote the feedback hub feedback to an MSFT: work item. That's weird, but I think we're changing some backend bits. 
* I can still get to feedback ID `9a191c39-d40a-4559-82ed-d23e3cfec3a1`
* It's got three `.dmp`s for `explorer.exe`, 
* and a link to a bucket: `1568601951875420617` (`BucketDetails?BucketIdentifier=1568601951875420617`). 
  * There's only exactly one hit for this specific bucket in the last 90 days. 
  * It's a `moapphang` in `microsoft.windowsterminal_1.19.10821.0_x64__8wekyb3d8bbwe`, and we have _other_ moapphang's hanging around too, but I don't think they're related
  * There's nothing else interesting in the bucket...

**Comment by zadjii-msft** [comment]
Dustin repro'd this instantaneously on Windows 10, and Leonard thinks... I think this code is the problem:

https://github.com/microsoft/terminal/blob/49e4eea60f737b46b8aeda505f4693df8a9d44a6/src/cascadia/WinRTUtils/inc/Utils.h#L15-L55

https://github.com/microsoft/terminal/blob/49e4eea60f737b46b8aeda505f4693df8a9d44a6/src/cascadia/TerminalSettingsEditor/Profiles_Base.cpp#L109-L125

Those sorta attempt to do the work async, but actually don't make it async at all

**Comment by KamenRevice** [comment]
Really? Did you locate the bug so quickly, and how long will it take to fix it, please?

**Comment by KamenRevice** [comment]
Really? Did you locate the bug so quickly, and how long will it take to fix it, please?

**Comment by zadjii-msft** [comment]
I have no idea if that's the actual source of the bug or not. We had coincidentally just been chatting about how this particular code didn't work as expected. And it sure _seems_ like this is the kind of thing that might cause this bug. 

No ideas how long it would take to fix. It might not even be the correct source of the bug 🤷

**Comment by katakura-yoichi** [comment]
Still reproduces on 1.24.11911.0 (and on Preview 1.25). I think I have a reliable repro key for this: **it only happens when a TSF-based CJK IME is the active input method.**

Environment: Windows 11, Japanese locale, default Microsoft IME.

I hit the same hang through a different entry point — the tab context menu's **Export Text** (`exportBuffer`), which goes through the same `SaveFilePicker` / `FilePicker` helper in `WinRTUtils/inc/Utils.h` that was suspected above.

Steps:
1. Tab context menu → Export Text. The Save dialog opens normally.
2. Click into the filename field (or the address bar) and press any key.
3. The entire Terminal UI hangs permanently. Only killable from Task Manager.

Observations that may help narrow it down:

- **Mouse-only interaction works fine.** Navigating folders, right-click → Paste into the filename field, and clicking Save all work. The hang is triggered by the first keystroke, not by focus.
- Switching the input language to English (US) with Win+Space *before* typing → no hang, works every time.
- Putting the Japanese IME into direct-input (英数) mode is **not** enough — it still hangs. Merely having the TSF text service attached is sufficient.
- Enabling "Use previous version of Microsoft IME" (which falls back to IMM32 instead of TSF) **also fixes it**.
- Notepad's Save As dialog in the same session is completely unaffected, so this is not a shell / common-dialog problem.
- Task Manager's "Analyze wait chain" reports `WindowsTerminal.exe` as running normally, i.e. it is not blocked on another process. The hang is inside Terminal's own UI thread.

This seems consistent with the analysis above: `FilePicker()` calls `fileDialog->Show(parentHwnd)` synchronously, so the dialog's modal loop ends up owning the XAML UI thread rather than the call actually being async. Terminal also runs its own TSF handler on that same thread for terminal IME input, and TSF appears to deadlock when the dialog's edit control tries to activate a text service on that blocked thread. IMM32 does not go through that path, which would explain why the legacy-IME workaround avoids it.

This may also be why it has been hard to reproduce: the original report's screenshot shows a Chinese-locale Windows, and I'm on Japanese, while an English-locale machine would likely never hit it.

Happy to capture a dump of the hung process if that would be useful.

**Comment by katakura-yoichi** [comment]
Follow-up: I captured a hang dump and analysed it. The dump itself is a full-memory dump of a session with live SSH connections, so I can't attach it here, but I can route it through Feedback Hub if that's useful — just say the word.

**The file dialog's modal loop is running nested inside the XAML UI thread's message loop.** Reconstructed call structure of the UI thread, outermost first:

```
WindowsTerminal.exe            (main / message loop)
  user32.dll                   (message dispatch)
    CoreMessaging.dll / Windows.UI.dll
      Windows.UI.Xaml.dll
        TerminalApp.dll        (Export Text handler -> SaveFilePicker)
          comdlg32.dll         (IFileDialog::Show)
            comdlg32.dll       (its own modal message loop)
              duser.dll / user32.dll
                win32u.dll     (message-wait syscall - where the thread is parked)
```

So `FilePicker()`'s `fileDialog->Show(parentHwnd)` really does block the XAML UI thread, which lines up with the suspicion in the comment above about that code not actually being async.

Two more things from the dump that may help:

- The UI thread is parked in a `MsgWaitForMultipleObjects`-style syscall — **not blocked on a lock and not blocked on another process**. Several idle worker threads in the same dump sit at the identical address, which is how I identified it. This is consistent with Task Manager's "Analyze wait chain" reporting nothing, and suggests an input/message-delivery stall rather than a classic deadlock.
- Input-stack modules loaded: `msctf.dll`, `TextInputFramework.dll`, `InputHost.dll`, `Windows.UI.Core.TextInput.dll`, `tiptsf.dll`. No IMM32-era IME DLL is present, consistent with the TSF-based Microsoft IME being active when the hang happened. Notably, **no thread has any `msctf` frame on its stack** — the IME isn't stuck inside a call, the UI thread just never gets what it's waiting for.

Environment: Terminal 1.24.11911.0, Windows 11 build 26200, x64, Japanese locale.

Caveat on method: I don't have private symbols, so this is a stack-memory scan resolved to module+offset rather than a proper symbolised unwind. The layering is unambiguous, but individual frames may include stale leftovers and I can't give you function names.

**Comment by unclebob2024** [comment]
I'm using **Microsoft Pinyin IME**. I also reported this issue in [#20426](https://github.com/microsoft/terminal/issues/20426)
